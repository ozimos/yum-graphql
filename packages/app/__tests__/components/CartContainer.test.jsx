/* global React:false, shallow:false toJson:false */

import { CartContainer } from
  '../../src/components/orderCart/ConnectedCartContainer';
import { Meals } from '../__mocks__/mealDataMock';

const props = {
  order: Meals,
  orderId: 'abc',
  orderError: '',
  dispatch: jest.fn(),
  notify: jest.fn(),
  MealRow: jest.fn(),
  clearCart: jest.fn(),
};

describe('CartContainer Component', () => {

  const setup = () => shallow(<CartContainer {...props} />);

  it('should render without throwing an error', () => {
    const wrapper = setup();

    expect(wrapper
      .exists(<table className="table" />)).toBe(true);
  });

  it('renders correctly', () => {
    let wrapper = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
    const propsWithExtraClass = { ...props, addClass: 'string' };
    const setupExtraClass = () => shallow(<CartContainer
      {...propsWithExtraClass}
    />);
    wrapper = setupExtraClass();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it(
    'should call `placeOrder` for new orders' +
    ' when the place order button is clicked',
    () => {
      const wrapper = setup();

      const placeOrderSpy = jest.spyOn(
        wrapper.instance(),
        'placeOrder'
      );
      wrapper.instance().placeOrder();
      expect(placeOrderSpy).toHaveBeenCalled();
    }
  );

  it('should call `placeOrder` for existing orders ' +
  'when the place order button is clicked', () => {
    const orderId = '';
    const propsWithOrderId = { ...props, orderId };
    const setupWithOrderId = () => shallow(<CartContainer
      {...propsWithOrderId}
    />);

    const wrapper = setupWithOrderId();

    const placeOrderSpy = jest.spyOn(
      wrapper.instance(),
      'placeOrder'
    );
    wrapper.instance().placeOrder();
    expect(placeOrderSpy).toHaveBeenCalled();
  });

  it('should call `setQuantity` when the meal quantity is changed' +
  ' to a negative value', () => {
    const wrapper = setup();

    const setQuantitySpy = jest.spyOn(
      wrapper.instance(),
      'setQuantity'
    );
    wrapper.instance().setQuantity({ target: { value: -1 } }, 'abc');
    expect(setQuantitySpy).toHaveBeenCalled();
  });

  it('should call `setQuantity` when the meal quantity is changed' +
  ' above max values', () => {
    const wrapper = setup();

    const setQuantitySpy = jest.spyOn(
      wrapper.instance(),
      'setQuantity'
    );
    wrapper.instance().setQuantity({ target: { value: 99 } }, 'abc');
    expect(setQuantitySpy).toHaveBeenCalled();
  });

  it('should call `setQuantity`when the meal quantity is changed ' +
  'to a value in the normal range', () => {
    const wrapper = setup();

    const setQuantitySpy = jest.spyOn(
      wrapper.instance(),
      'setQuantity'
    );
    wrapper.instance().setQuantity({ target: { value: 20 } }, 'abc');
    expect(setQuantitySpy).toHaveBeenCalled();
  });
});

