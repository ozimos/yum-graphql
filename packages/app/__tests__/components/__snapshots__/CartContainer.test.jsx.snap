// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartContainer Component renders correctly 1`] = `
<div
  className=""
  id="order-cart"
>
  <div
    className="flexbox cart"
    id="order-cart-heading"
  >
    <h5>
      Order Cart
    </h5>
    <button
      className="btn title-button"
    >
      ✖
    </button>
  </div>
  <div
    className="responsive-table"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <th>
            Meal Title
          </th>
          <th>
            Price (₦)
          </th>
          <th>
            Quantity
          </th>
          <th>
            Subtotal (₦)
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <mockConstructor
          addClass=""
          clearCart={[MockFunction]}
          description="nice"
          dispatch={[MockFunction]}
          id="meal1"
          imageUrl="nice"
          key="meal1"
          notify={[MockFunction]}
          orderError=""
          orderId="abc"
          price={1200}
          quantity={1}
          setQuantity={[Function]}
          title="Eba and Egusi"
          userId="user1"
        />
        <mockConstructor
          addClass=""
          clearCart={[MockFunction]}
          description="nice"
          dispatch={[MockFunction]}
          id="meal2"
          imageUrl="nice"
          key="meal2"
          notify={[MockFunction]}
          orderError=""
          orderId="abc"
          price={2000}
          quantity={1}
          setQuantity={[Function]}
          title="Rice and Stew"
          userId="user1"
        />
      </tbody>
      <caption
        className="flexbox-cart"
      >
        <div
          className="flexbox"
        >
          <div
            className="flexbox info"
          >
            <p>
               Order Total 
            </p>
            <p>
               ₦
              3200
            </p>
          </div>
          <div
            className="flexbox info"
          >
            <button
              className="btn btn-cart"
              disabled={false}
              id="place-order"
              onClick={[Function]}
            >
              Modify Order
            </button>
            <button
              className="btn btn-cart"
              disabled={false}
              onClick={[MockFunction]}
            >
              Clear Cart
            </button>
          </div>
        </div>
      </caption>
    </table>
  </div>
</div>
`;

exports[`CartContainer Component renders correctly 2`] = `
<div
  className="string"
  id="order-cart"
>
  <div
    className="flexbox cart"
    id="order-cart-heading"
  >
    <h5>
      Order Cart
    </h5>
    <button
      className="btn title-button"
    >
      ✖
    </button>
  </div>
  <div
    className="responsive-table"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <th>
            Meal Title
          </th>
          <th>
            Price (₦)
          </th>
          <th>
            Quantity
          </th>
          <th>
            Subtotal (₦)
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <mockConstructor
          addClass="string"
          clearCart={[MockFunction]}
          description="nice"
          dispatch={[MockFunction]}
          id="meal1"
          imageUrl="nice"
          key="meal1"
          notify={[MockFunction]}
          orderError=""
          orderId="abc"
          price={1200}
          quantity={1}
          setQuantity={[Function]}
          title="Eba and Egusi"
          userId="user1"
        />
        <mockConstructor
          addClass="string"
          clearCart={[MockFunction]}
          description="nice"
          dispatch={[MockFunction]}
          id="meal2"
          imageUrl="nice"
          key="meal2"
          notify={[MockFunction]}
          orderError=""
          orderId="abc"
          price={2000}
          quantity={1}
          setQuantity={[Function]}
          title="Rice and Stew"
          userId="user1"
        />
      </tbody>
      <caption
        className="flexbox-cart"
      >
        <div
          className="flexbox"
        >
          <div
            className="flexbox info"
          >
            <p>
               Order Total 
            </p>
            <p>
               ₦
              3200
            </p>
          </div>
          <div
            className="flexbox info"
          >
            <button
              className="btn btn-cart"
              disabled={false}
              id="place-order"
              onClick={[Function]}
            >
              Modify Order
            </button>
            <button
              className="btn btn-cart"
              disabled={false}
              onClick={[MockFunction]}
            >
              Clear Cart
            </button>
          </div>
        </div>
      </caption>
    </table>
  </div>
</div>
`;
